### PR TITLE
CMR-9776: Resolve linter errors & warnings in common-app-lib; add linting config

### PR DESCRIPTION
**Pull request by zimzoom**
# Overview

### What is the feature/fix?

Fix clj-kondo errors and warnings on common-app-lib.

Also added .clj-kondo to codebase, including config file and imported config from clj-rewrite.

### What is the Solution?

See above, followed clj-kondo docs. See comments in config.edn file.

### What areas of the application does this impact?

Quality control

# Checklist

- [x ] I have updated/added unit and integration tests that prove my fix is effective or that my feature works
- [x ] New and existing unit and int tests pass locally and remotely with my changes
- [x ] I have removed unnecessary/dead code and imports in files I have changed
- [x ] I have performed a self-review of my own code
- [x ] I have commented my code, particularly in hard-to-understand areas
- [x ] I have made corresponding changes to the documentation
- [x ] My changes generate no new warnings

### CODE REVIEW

File: .clj-kondo/config.edn
Comment (eereiter) [inline .clj-kondo/config.edn:10]: This is much better than the declare statements that I was using.  
Can we also add are2 and are?  They are deprecated functions, but the test code still uses them and are showing up as such from the linters.

Couple of questions 
1) When you merge this in does this work when running kondo only from the top level of the project or can kondo be run inside of each project and it will pick up this file?  Or do you have to tell it about the file?  I am using my own installed cli-kondo executable at the moment.
2) After you merge in and I understand the answer to my first question I am assuming that I should remove the declare statements that I put in?

File: .clj-kondo/rewrite-clj/rewrite-clj/config.edn
Comment (eereiter) [inline .clj-kondo/rewrite-clj/rewrite-clj/config.edn:1]: What does this file do?

File: common-app-lib/src/cmr/common_app/api/request_context_user_augmenter.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/api/request_context_user_augmenter.clj:87]: You don't need the let statement.  Can you please remove it.

File: common-app-lib/src/cmr/common_app/api/routes.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/api/routes.clj:108]: Just remove _params and _headers, they are not used.

File: common-app-lib/src/cmr/common_app/api/routes.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/api/routes.clj:114]: You should be able to get rid of ":as _params" as it is not used.

File: common-app-lib/src/cmr/common_app/api/routes.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/api/routes.clj:116]: Get rid of _headers :headers as it is not used.

File: common-app-lib/src/cmr/common_app/api/routes.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/api/routes.clj:129]: Get rid of _headers :headers as it is not used.

File: common-app-lib/src/cmr/common_app/api/routes.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/api/routes.clj:144]: Get rid of _params and _headers as they are not used.

File: common-app-lib/src/cmr/common_app/api/routes.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/api/routes.clj:158]: Get rid of _params and _headers as they are not used.

File: common-app-lib/src/cmr/common_app/api/routes.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/api/routes.clj:164]: Get rid of _params and _headers as they are not used.

File: common-app-lib/src/cmr/common_app/app_events.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/app_events.clj:57]: Get rid of _app-name in this function and then remove (:instance-name system) from the following files:
ingest-app/src/cmr/ingest/runner.clj
search-app/src/cmr/search/runner.clj 

from the stop-on-exit-hook function

File: common-app-lib/src/cmr/common_app/humanizer.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/humanizer.clj:30]: I have been converting the defmulti/defmethods into a function with case statement.  We don't need polymorphism here especially since all of the to-human functions are in this namespace.  Polymorphism is a little slower with regard to performance of the software.
But I will let you decide if you want to keep this as is or convert it.  I am not going to comment on the defmethod functions below.

File: common-app-lib/src/cmr/common_app/services/kms_fetcher.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/kms_fetcher.clj:22]: Are you sure you don't need the import for the exception?  Usually these have been put into the namespaces to avoid the class not found exception. when the compiler removes the ExceptionInfo since it thinks that it isn't being used.

File: common-app-lib/src/cmr/common_app/services/kms_lookup.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/kms_lookup.clj:32]: Are you sure you don't need the import for the exception? Usually these have been put into the namespaces to avoid the class not found exception. when the compiler removes the ExceptionInfo since it thinks that it isn't being used.

File: common-app-lib/src/cmr/common_app/services/search.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/search.clj:65]: I have been converting the defmulti/defmethod functions to a function with a case statement when possible (it is almost always possible since most of the time all of the defmethod functions are in the same namespace), but I will let you decide if you want to do that.

File: common-app-lib/src/cmr/common_app/services/search.clj
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/search.clj:71]: I have been converting the defmulti/defmethod functions to a function with a case statement when possible (it is almost always possible since most of the time all of the defmethod functions are in the same namespace), but I will let you decide if you want to do that.

**Review by eereiter — CHANGES_REQUESTED** [review]
There are some cases where you can purge a little more code than just use _ in front of a parameter.

File: .clj-kondo/config.edn
Comment (zimzoom) [reply .clj-kondo/config.edn:10]: 1. Interestingly, I would have thought it would only work from the root level as well, but I just tried it and it does still work in the subprojects. I am using my own clj-kondo executable as well and have not added clj-kondo as a dep to CMR, just this config file since clj-kondo is now our officially supported linter to my knowledge.

I tested this by changing into the common-app-lib directory and running `clj-kondo --lint ./src`, which still suppressed all the stuff mentioned here. To make sure I tried deleting the contents of config.edn and running it again and it did come back.

2. I think that a lot of those, that you previously added, might be covered by what is here, but possibly not all. Maybe we should coordinate on rules for what to consistently ignore/exclude/include etc.

Comment (zimzoom) [reply .clj-kondo/config.edn:10]: And yes, I will add are2 and are3 as well.

File: .clj-kondo/rewrite-clj/rewrite-clj/config.edn
Comment (zimzoom) [reply .clj-kondo/rewrite-clj/rewrite-clj/config.edn:1]: This file was pulled in automatically by following the instructions in clj-kondo's project set up guide, which is here: https://github.com/clj-kondo/clj-kondo?tab=readme-ov-file#project-setup

To my understanding, this file (and the folder it is in) was written by the authors of this package clj-rewrite, which we have as a dependency, and they "exported" it so that projects using their package can import it and thus lint clj-rewrite correctly.

This seems to be the only one that exported a clj-kondo config. The super common ones like compojure are included in clj-kondo by default. The docs recommended adding this to version control. (see link above)

File: common-app-lib/src/cmr/common_app/api/request_context_user_augmenter.clj
Comment (zimzoom) [reply common-app-lib/src/cmr/common_app/api/request_context_user_augmenter.clj:87]: Ok, I see I was too conservative, fixing all these now. Thank you

File: .clj-kondo/rewrite-clj/rewrite-clj/config.edn
Comment (zimzoom) [reply .clj-kondo/rewrite-clj/rewrite-clj/config.edn:1]: Originally I wanted to incorporate this into the build, but when I saw only 1 dep came out of it, I figured it wasn't worth it.

File: .clj-kondo/config.edn
Comment (eereiter) [reply .clj-kondo/config.edn:10]: it is just are2. are3 you already included and are is a clojure command.

Comment (zimzoom) [reply .clj-kondo/config.edn:10]: Note for Reader: discussed with reviewer here and we decided I will review the aforementioned declares that were already merged in for access-control-app, acl-lib, and es-spatial-plugin, and incorporate into this PR.

File: common-app-lib/src/cmr/common_app/humanizer.clj
Comment (zimzoom) [reply common-app-lib/src/cmr/common_app/humanizer.clj:30]: Yes, I competely agree!

File: common-app-lib/src/cmr/common_app/services/search.clj
Comment (zimzoom) [reply common-app-lib/src/cmr/common_app/services/search.clj:71]: In this particular case, there do exist defmethods in other namespaces. Results handler namespaces to be exact, which account for a lot of the scattered multimethods.

File: common-app-lib/src/cmr/common_app/app_events.clj
Comment (jceaser) [reply common-app-lib/src/cmr/common_app/app_events.clj:57]: I think I added those so that other apps could print out the name of themselves as they were being killed.

File: common-app-lib/src/cmr/common_app/services/kms_fetcher.clj
Comment (jceaser) [reply common-app-lib/src/cmr/common_app/services/kms_fetcher.clj:22]: when is this error message generated, at run time or compile time?

Comment (zimzoom) [reply common-app-lib/src/cmr/common_app/services/kms_fetcher.clj:22]: Resolving because this was undone but git is not showing as Outdated

File: common-app-lib/src/cmr/common_app/services/kms_lookup.clj
Comment (zimzoom) [reply common-app-lib/src/cmr/common_app/services/kms_lookup.clj:32]: Resolving because this was undone but git is not showing as Outdated

File: es-spatial-plugin/project.clj
Comment (jceaser) [inline es-spatial-plugin/project.clj:1]: what is the error here? which namespace can it not find?

File: ingest-app/src/cmr/ingest/runner.clj
Comment (jceaser) [inline ingest-app/src/cmr/ingest/runner.clj:13]: so the system name was to print out which task in AWS closed successfully. I was thinking that we could track how many started and never report a clean shutdown.

File: umm-spec-lib/src/cmr/umm_spec/umm_to_xml_mappings/dif9/data_contact.clj
Comment (jceaser) [inline umm-spec-lib/src/cmr/umm_spec/umm_to_xml_mappings/dif9/data_contact.clj:3]: don't we still want to not use the refer all?

File: umm-spec-lib/src/cmr/umm_spec/umm_to_xml_mappings/iso19115_2/additional_attribute.clj
Comment (jceaser) [inline umm-spec-lib/src/cmr/umm_spec/umm_to_xml_mappings/iso19115_2/additional_attribute.clj:4]: same, don't we want to not do refer all?

File: umm-spec-lib/src/cmr/umm_spec/umm_to_xml_mappings/iso19115_2/metadata_association.clj
Comment (jceaser) [inline umm-spec-lib/src/cmr/umm_spec/umm_to_xml_mappings/iso19115_2/metadata_association.clj:4]: refer all bad?

File: common-app-lib/src/cmr/common_app/app_events.clj
Comment (zimzoom) [reply common-app-lib/src/cmr/common_app/app_events.clj]: Good point Thomas, we missed the purpose/the supporter here. Reverting change

File: ingest-app/src/cmr/ingest/runner.clj
Comment (zimzoom) [reply ingest-app/src/cmr/ingest/runner.clj:13]: Good point, we missed the purpose/the supporter. Reverting change

File: es-spatial-plugin/project.clj
Comment (jceaser) [reply es-spatial-plugin/project.clj:1]: this is a project file, lets leave this here so that we know what kondo is all about.

File: umm-spec-lib/src/cmr/umm_spec/url.clj
Comment (TylerHeald1) [inline umm-spec-lib/src/cmr/umm_spec/url.clj:10]: This doesn't seem equivalent. This type hint is being applied to the input arg rather than the function, isn't it?

Comment (zimzoom) [reply umm-spec-lib/src/cmr/umm_spec/url.clj:10]: Oops, you're correct. My mistake; correcting

**Review by eereiter — APPROVED** [review]

**Review by jceaser — APPROVED** [review]

File: .clj-kondo/config.edn
Comment (jceaser) [reply .clj-kondo/config.edn]: reader has noted discussion.

**Pull request merged**
